pkg/storage/mongo/mongo.go

```diff
@@ -27,69 +28,70 @@ type ModuleStore struct {
 
 // NewStorage returns a connected Mongo backed storage
 // that satisfies the Backend interface.
-func NewStorage(conf *config.MongoConfig, timeout time.Duration) (*ModuleStore, error) {
+func NewStorage(conf *config.MongoConfig, timeout time.Duration, insecure bool) (*ModuleStore, error) {
 	const op errors.Op = "mongo.NewStorage"
 	if conf == nil {
 		return nil, errors.E(op, "No Mongo Configuration provided")
 	}
-	ms := &ModuleStore{url: conf.URL, certPath: conf.CertPath, timeout: timeout}
+	ms := &ModuleStore{url: conf.URL, certPath: conf.CertPath, timeout: timeout, insecure: insecure}
+	client, err := ms.newClient(conf)
+	ms.client = client
+
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+
+	_, err = ms.connect(conf)
 
-	err := ms.connect(conf)
 	if err != nil {
 		return nil, errors.E(op, err)
 	}
 
 	return ms, nil
 }
 
-func (m *ModuleStore) connect(conf *config.MongoConfig) error {
+func (m *ModuleStore) connect(conf *config.MongoConfig) (*mongo.Collection, error) {
 	const op errors.Op = "mongo.connect"
 
 	var err error
-	m.s, err = m.newSession(m.timeout, m.insecure, conf)
+	err = m.client.Connect(context.Background())
 
 	if err != nil {
-		return errors.E(op, err)
+		return nil, errors.E(op, err)
 	}
 
-	return m.initDatabase()
+	return m.initDatabase(), nil
 }
 
-func (m *ModuleStore) initDatabase() error {
-	m.c = "modules"
-
-	index := mgo.Index{
-		Key:        []string{"base_url", "module", "version"},
-		Unique:     true,
-		Background: true,
-		Sparse:     true,
-	}
-	c := m.s.DB(m.d).C(m.c)
-	return c.EnsureIndex(index)
+func (m *ModuleStore) initDatabase() *mongo.Collection {
+	// TODO: database and collection as env vars, or params to New()? together with user/mongo
```

Comment:
I think it's ok to leave this TODO out. We've hard-coded the db and collection names since the first version and nobody has asked to change them so far, so we can wait and see if someone does...